tensorflow/java/src/main/java/org/tensorflow/Scope.java

```diff
@@ -0,0 +1,198 @@
+   * @return a new subscope
+   * @throws IllegalArgumentException if the name is invalid
+   */
+  public Scope withSubScope(String childScopeName) {
+    return toBuilder().nameScope(nameScope.withSubScope(childScopeName)).build();
+  }
+
+  /**
+   * Return a new scope that uses the provided name for an op.
+   *
+   * <p>Operations created within this scope will have a name of the form {@code
+   * name/opName[_suffix]}. This lets you name a specific operator more meaningfully.
+   *
+   * <p>Valid operator names must match the regular expression {@code [A-Za-z0-9.][A-Za-z0-9_.\\-]*}
+   *
+   * @param opName name for an operator in the returned scope
+   * @return a new Scope that uses opName for operations.
+   * @throws IllegalArgumentException if the name is invalid
+   */
+  public Scope withOpName(String opName) {
+    return toBuilder().nameScope(nameScope.withOpName(opName)).build();
+  }
+
+  /**
+   * Create a unique name for an operator, using a provided default if necessary.
+   *
+   * <p>This is normally called only by operator building classes.
+   *
+   * <p>This method generates a unique name, appropriate for the name scope controlled by this
+   * instance. Typical operator building code might look like
+   *
+   * <pre>{@code
+   * scope.graph().opBuilder("Constant", scope.makeOpName("Constant"))...
+   * }</pre>
+   *
+   * <p><b>Note:</b> if you provide a composite operator building class (i.e, a class that adds a
+   * set of related operations to the graph by calling other operator building code) you should also
+   * create a {@link #withSubScope(String)} scope for the underlying operators to group them under a
+   * meaningful name.
+   *
+   * <pre>{@code
+   * public static Stddev create(Scope scope, ...) {
+   *   // group sub-operations under a common name
+   *   Scope group = scope.withSubScope("stddev");
+   *   ... Sqrt.create(group, Mean.create(group, ...))
+   * }
+   * }</pre>
+   *
+   * @param defaultName name for the underlying operator.
+   * @return unique name for the operator.
+   * @throws IllegalArgumentException if the default name is invalid.
+   */
+  public String makeOpName(String defaultName) {
+    return nameScope.makeOpName(defaultName);
+  }
+
+  private Scope(Builder builder) {
+    graph = builder.graph;
+    if (builder.nameScope != null) {
+      nameScope = builder.nameScope;
+    } else {
+      nameScope = NameScope.create();
+    }
+  }
+
+  private final Graph graph;
+  private final NameScope nameScope;
+
+  private Builder toBuilder() {
+    return builder(graph).nameScope(nameScope);
+  }
+
+  private static Builder builder(Graph graph) {
+    return new Builder(graph);
+  }
+
+  private static final class Builder {
+    private Builder(Graph g) {
+      graph = g;
+    }
+
+    private Builder nameScope(NameScope ns) {
+      nameScope = ns;
+      return this;
+    }
+
+    private Scope build() {
+      return new Scope(this);
```

Comment:
That sounds great :-) Your summary about helper scopes is just how we're thinking about it as well. I'll update the PR with all the changes you suggested (but keep the helper scopes as-is) and go from there. Thanks!